src/cargo/util/lints.rs

```diff
@@ -0,0 +1,241 @@
+            Some((_, toml_lint)) => toml_lint.level().into(),
+            None => {
+                if let Some((lint_edition, lint_level)) = self.edition_lint_opts {
+                    if edition >= lint_edition {
+                        lint_level
+                    } else {
+                        IMPLICIT_FEATURES.default_level
+                    }
+                } else {
+                    IMPLICIT_FEATURES.default_level
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum LintLevel {
+    Allow,
+    Warn,
+    Deny,
+    Forbid,
+}
+
+impl LintLevel {
+    pub fn to_diagnostic_level(self) -> Level {
+        match self {
+            LintLevel::Allow => Level::Note,
+            LintLevel::Warn => Level::Warning,
+            LintLevel::Deny => Level::Error,
+            LintLevel::Forbid => Level::Error,
+        }
+    }
+}
+
+impl Into<LintLevel> for TomlLintLevel {
+    fn into(self) -> LintLevel {
+        match self {
+            TomlLintLevel::Allow => LintLevel::Allow,
+            TomlLintLevel::Warn => LintLevel::Warn,
+            TomlLintLevel::Deny => LintLevel::Deny,
+            TomlLintLevel::Forbid => LintLevel::Forbid,
+        }
+    }
+}
+
+const IMPLICIT_FEATURES: Lint = Lint {
+    name: "implicit-features",
+    desc: "warn about the use of unstable features",
+    groups: &[RUST_2024_COMPATIBILITY],
+    default_level: LintLevel::Allow,
+    edition_lint_opts: Some((Edition::Edition2024, LintLevel::Deny)),
+};
+
+pub fn check_implicit_features(
+    pkg: &Package,
+    path: &Path,
+    lints: &TomlToolLints,
+    error_count: &mut usize,
+    gctx: &GlobalContext,
+) -> CargoResult<()> {
+    let lint_level = IMPLICIT_FEATURES.level(lints, pkg.manifest().edition());
+    if lint_level == LintLevel::Allow {
+        return Ok(());
+    }
+
+    let manifest = pkg.manifest();
+    let features: BTreeMap<InternedString, Vec<InternedString>> = manifest
+        .resolved_toml()
+        .features()
+        .map_or(BTreeMap::new(), |f| {
+            f.iter()
+                .map(|(k, v)| {
+                    (
+                        InternedString::new(k),
+                        v.iter().map(InternedString::from).collect(),
+                    )
+                })
+                .collect()
+        });
+    let map: FeatureMap = features
+        .iter()
+        .map(|(feature, list)| {
+            let fvs: Vec<_> = list
+                .iter()
+                .map(|feat_value| FeatureValue::new(*feat_value))
+                .collect();
+            (*feature, fvs)
+        })
+        .collect();
+
+    // Add implicit features for optional dependencies if they weren't
+    // explicitly listed anywhere.
+    let explicitly_listed: HashSet<_> = map
+        .values()
+        .flatten()
+        .filter_map(|fv| match fv {
+            Dep { dep_name } => Some(*dep_name),
+            _ => None,
+        })
+        .collect();
```

Comment:
Two hash sets are good enough?

```suggestion
    let manifest = pkg.manifest();
    let user_defined_features = manifest.resolved_toml().features();
    let features = user_defined_features.map_or(HashSet::new(), |f| {
        f.keys().map(|k| InternedString::new(&k)).collect()
    });
    // Add implicit features for optional dependencies if they weren't
    // explicitly listed anywhere.
    let explicitly_listed = user_defined_features.map_or(HashSet::new(), |f| {
        f.values()
            .flatten()
            .filter_map(|v| match FeatureValue::new(v.into()) {
                Dep { dep_name } => Some(dep_name),
                _ => None,
            })
            .collect()
    });
```